d2ir/d2ir.go

```diff
@@ -0,0 +1,1044 @@
+	for i := len(f.References) - 1; i >= 0; i-- {
+		if f.References[i].Primary() {
+			return f.References[i]
+		}
+	}
+	return nil
+}
+
+func (f *Field) LastPrimaryKey() *d2ast.Key {
+	fr := f.lastPrimaryRef()
+	if fr == nil {
+		return nil
+	}
+	return fr.Context.Key
+}
+
+func (f *Field) LastRef() Reference {
+	return f.References[len(f.References)-1]
+}
+
+type EdgeID struct {
+	SrcPath  []string `json:"src_path"`
+	SrcArrow bool     `json:"src_arrow"`
+
+	DstPath  []string `json:"dst_path"`
+	DstArrow bool     `json:"dst_arrow"`
+
+	// If nil, then any EdgeID with equal src/dst/arrows matches.
+	Index *int `json:"index"`
+}
+
+func NewEdgeIDs(k *d2ast.Key) (eida []*EdgeID) {
+	for _, ke := range k.Edges {
+		eid := &EdgeID{
+			SrcPath:  ke.Src.IDA(),
+			SrcArrow: ke.SrcArrow == "<",
+			DstPath:  ke.Dst.IDA(),
+			DstArrow: ke.DstArrow == ">",
+		}
+		if k.EdgeIndex != nil {
+			eid.Index = k.EdgeIndex.Int
+		}
+		eida = append(eida, eid)
+	}
+	return eida
+}
+
+func (eid *EdgeID) Copy() *EdgeID {
+	tmp := *eid
+	eid = &tmp
+
+	eid.SrcPath = append([]string(nil), eid.SrcPath...)
+	eid.DstPath = append([]string(nil), eid.DstPath...)
+	return eid
+}
+
+func (eid *EdgeID) Match(eid2 *EdgeID) bool {
+	if eid.Index != nil && eid2.Index != nil {
+		if *eid.Index != *eid2.Index {
+			return false
+		}
+	}
+
+	if len(eid.SrcPath) != len(eid2.SrcPath) {
+		return false
+	}
+	if eid.SrcArrow != eid2.SrcArrow {
+		return false
+	}
+	for i, s := range eid.SrcPath {
+		if !strings.EqualFold(s, eid2.SrcPath[i]) {
+			return false
+		}
+	}
+
+	if len(eid.DstPath) != len(eid2.DstPath) {
+		return false
+	}
+	if eid.DstArrow != eid2.DstArrow {
+		return false
+	}
+	for i, s := range eid.DstPath {
+		if !strings.EqualFold(s, eid2.DstPath[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (eid *EdgeID) resolveUnderscores(m *Map) (*EdgeID, *Map, error) {
+	eid = eid.Copy()
+	maxUnderscores := go2.Max(countUnderscores(eid.SrcPath), countUnderscores(eid.DstPath))
+	for i := 0; i < maxUnderscores; i++ {
+		if eid.SrcPath[0] == "_" {
+			eid.SrcPath = eid.SrcPath[1:]
+		} else {
+			mf := ParentField(m)
+			eid.SrcPath = append([]string{mf.Name}, eid.SrcPath...)
+		}
+		if eid.DstPath[0] == "_" {
+			eid.DstPath = eid.DstPath[1:]
+		} else {
+			mf := ParentField(m)
+			eid.DstPath = append([]string{mf.Name}, eid.DstPath...)
+		}
+		m = ParentMap(m)
+		if m == nil {
+			return nil, nil, errors.New("invalid underscore")
+		}
+	}
+	return eid, m, nil
+}
+
+func (eid *EdgeID) trimCommon() (common []string, _ *EdgeID) {
+	eid = eid.Copy()
+	for len(eid.SrcPath) > 1 && len(eid.DstPath) > 1 {
+		if !strings.EqualFold(eid.SrcPath[0], eid.DstPath[0]) {
+			return common, eid
+		}
+		common = append(common, eid.SrcPath[0])
+		eid.SrcPath = eid.SrcPath[1:]
+		eid.DstPath = eid.DstPath[1:]
+	}
+	return common, eid
+}
+
+type Edge struct {
+	// *Map
+	parent Node
+
+	ID *EdgeID `json:"edge_id"`
+
+	Primary_ *Scalar `json:"primary,omitempty"`
```

Comment:
Nothing, just there's a method on each `Node` `Primary()` and I still wanted this to be exported so that it's encoded by `encoding/json` and so I couldn't prefix `_`.

Reply:
I'll add a comment.